src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,233 @@
+    } while (i != hashValue);
+
+    if (!found)
+    {
+        // Remove one randomly
+        i = GetRandU16() % kServicePoolCapacity;
+    }
+
+    if (mEntries[i].mLazyDelete)
+    {
+        mLazyDeleteCount--;
+        mEntries[i].mLazyDelete = false;
+    }
+    mEntries[i].mService  = service;
+    mEntries[i].mNodeId   = nodeId;
+    mEntries[i].mFabricId = fabricId;
+    if (deepCopy)
+    {
+        error = CopyTextEntries(&mEntries[i].mService, service);
+        if (error != CHIP_NO_ERROR)
+        {
+            mEntries[i].mNodeId = kUndefinedNodeId;
+        }
+    }
+    return error;
+}
+
+CHIP_ERROR ServicePool::RemoveService(uint64_t nodeId, uint64_t fabricId)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    if (nodeId == kUndefinedNodeId)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    do
+    {
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            break;
+        }
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            mEntries[i].mLazyDelete = true;
+            mLazyDeleteCount++;
+            FreeTextEntries(&mEntries[i].mService);
+        }
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ServicePool::UpdateService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service)
+{
+    MdnsService * foundService;
+
+    if (!FindService(nodeId, fabricId, &foundService))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    FreeTextEntries(foundService);
+    *foundService = service;
+    return CopyTextEntries(foundService, service);
+}
+
+bool ServicePool::FindService(uint64_t nodeId, uint64_t fabricId, MdnsService ** service)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    if (nodeId == kUndefinedNodeId)
+    {
+        *service = nullptr;
+        return false;
+    }
+
+    do
+    {
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            break;
+        }
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            *service = &mEntries[i].mService;
+            return true;
+        }
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    *service = nullptr;
+    return false;
+}
+
+void ServicePool::Clear()
+{
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
+    {
+        mEntries[i].mNodeId     = kUndefinedNodeId;
+        mEntries[i].mLazyDelete = false;
+    }
+    mLazyDeleteCount = 0;
+}
+
+void ServicePool::ReHash()
+{
+    ServicePoolEntry copyEntries[kServicePoolCapacity];
+
+    memcpy(&copyEntries, mEntries, sizeof(mEntries));
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
```

Comment:
should there be some TODO here and below? these are loops that do nothing.

Reply:
Debug loops, removed.